Source/Samples/113_Actions/ActionDemo.cpp

```diff
@@ -0,0 +1,162 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+#include <Urho3D/Input/Input.h>
+#include <Urho3D/UI/Font.h>
+#include <Urho3D/UI/Text.h>
+#include <Urho3D/UI/UI.h>
+#include <Urho3D/UI/UIEvents.h>
+#include <Urho3D/Actions/ActionBuilder.h>
+#include <Urho3D/Actions/ActionManager.h>
+
+#include "ActionDemo.h"
+
+#include <Urho3D/DebugNew.h>
+
+ActionDemo::ActionDemo(Context* context)
+    : Sample(context)
+{
+    // Set the mouse mode to use in the sample
+    SetMouseMode(MM_FREE);
+    SetMouseVisible(true);
+}
+
+void ActionDemo::Start()
+{
+    // Execute base class startup
+    Sample::Start();
+
+    // Create UI elements
+    CreateUI();
+
+    // Finally subscribe to the update event. Note that by subscribing events at this point we have already missed some
+    // events like the ScreenMode event sent by the Graphics subsystem when opening the application window. To catch
+    // those as well we could subscribe in the constructor instead.
+    SubscribeToEvents();
+}
+
+ActionDemo::DemoElement& ActionDemo::AddElement(const IntVector2& pos, Actions::BaseAction* action)
+{
+    auto uiRoot = GetUIRoot();
+    auto& element = markers_.push_back();
+    element.element_ = uiRoot->CreateChild<Sprite>();
+    element.element_->SetEnabled(true);
+    element.element_->SetSize(20, 20);
+    element.element_->SetPosition(pos);
+    element.action_ = action;
+    return element;
+}
+
+void ActionDemo::CreateUI()
+{
+    auto* cache = GetSubsystem<ResourceCache>();
+    auto* font = cache->GetResource<Font>("Fonts/Anonymous Pro.ttf");
+
+    auto uiRoot = GetUIRoot();
+
+    // Construct new Text object
+    SharedPtr<Text> helloText(MakeShared<Text>(context_));
+
+    // Set String to display
+
+    helloText->SetText("Click on quads to trigger actions");
+
+    // Set font and text color
+    helloText->SetFont(font, 15);
+
+    // Align Text center-screen
+    helloText->SetHorizontalAlignment(HA_CENTER);
+    helloText->SetVerticalAlignment(VA_TOP);
+
+    // Add Text instance to the UI root element
+    uiRoot->AddChild(helloText);
+
+    IntVector2 pos{100, 64};
+    IntVector2 step{32, 0};
+    Vector2 offset{0.0f, 100.0f};
+    AddElement(pos, ActionBuilder(context_).MoveBy2D(1.0f, offset).DelayTime(2.0f).JumpBy2D(-offset).Build());
```

Comment:
Use-after-free bug here. `AddElement()` should take `SharedPtr<BaseAction>`, because that is what `Build()` returns.

Reply:
I do remember you've told me years ago that in Urho3D interface should take objects by pointer, not by smart pointer :) Maybe I misunderstood you.

Reply:
Well yes. General rule is this: use SharedPtr only when ownership is transferred (factory is an example). So in this case it kind of is fine